Pull request for #59: Fix size warnings

M-R-Sullivan opened the pull request:
For #59.

M-R-Sullivan commented:
@lboxell - thanks for PRing this. I merged master into dev_four with [this commit](https://github.com/gslab-econ/gslab_python/commit/b1d0ffffef66110326405d7b3cd17073e2a30391); it may be a good idea for you to review the ranges I made there, too.

lboxell commented:
Note: I changed base back-and-forth to refresh the file change comparisons.

lboxell commented on gslab_scons/_release_tools.py:
```diff
@@ -238,6 +238,70 @@ def up_to_date(mode = 'scons', directory = '.'):
+    ignore_dirs  = []
+    ignore_files = []
+    for line in message[ignored_line:len(message)]:
+        # Determine if line could be in path's directory
```
I don't quite follow this comment. The phrase "path's directory" is unclear to me.

M-R-Sullivan replied:
I removed this comment and the code beneath it in light of my removal of the directory argument.

lboxell commented on gslab_scons/_release_tools.py:
```diff
@@ -238,6 +238,70 @@ def up_to_date(mode = 'scons', directory = '.'):
+    '''List files ignored by git 
+
+    This function returns a list of the files in the directory
+    given by the dir_path argument that are ignored by git. 
```
I'm not sure it's really necessary for this function to restrict attention to the files in a single directory. I think it would suffice to just get a normalized list of all files that are ignored and then if a file is in this ignore list, we don't issue a size warning.

M-R-Sullivan replied:
Removed directory

lboxell commented on gslab_scons/_release_tools.py:
```diff
@@ -238,6 +238,70 @@ def up_to_date(mode = 'scons', directory = '.'):
+        raise ReleaseError("The path argument does not specify an "
+                           "existing directory.")
+
+    for root, directories, files in os.walk(path):
```
`directories` is never used?

M-R-Sullivan replied:
Replaced with _

lboxell commented on gslab_scons/release.py:
```diff
@@ -97,10 +45,61 @@
+    versioned = dict()
+
+    for file_name in release.keys():
+        versioned[file_name] = release[file_name]
```
The creation of `versioned` seems unnecessary to me. It seems like it contains the exact same information as `release`.

M-R-Sullivan replied:
Good catch. I think the distinction was once relevant but that I forgot to update the code when it became irrelevant.

lboxell commented on gslab_scons/release.py:
```diff
@@ -97,10 +45,61 @@
+
+
+def issue_size_warnings(file_MB_limit, total_MB_limit, bytes_in_MB):
+    '''Issue warnings if the files versioned in release are too large'''
```
I'm a little confused about the role of this function. At this point, everything in release that is going to be versioned is already versioned? Really, we want these warnings to be printed at the `git commit` stage, which is before the `release` stage. Nothing new can actually be versioned in the release stage can it?

M-R-Sullivan replied:
Everything that is going to be versioned has been versioned, yes. I think the idea is to remind the user that they have probably done something wrong since the last release, which may be better than never issuing size warnings. I agree it would be better, in principle, to perform this check at the commit-stage. When I originally introduced the size warnings, I was mostly following [this comment](https://github.com/gslab-econ/template/issues/19#issuecomment-262024738) and some other notes in its issue.

lboxell replied:
What about porting this functionality to be run as the last step in scons? So we run scons and populate release, and at the last moment, we spit out a warning regarding file size. I think that would make it closer to the decision point of whether to commit large files or not.

lboxell commented on gslab_scons/release.py:
```diff
@@ -1,70 +1,19 @@
-                                create_size_dictionary,
-                                extract_dot_git, 
-                                release)
+import _release_tools as rtools
```
rtools sounds like it relates to R.

M-R-Sullivan replied:
Dropped `as rtools`

M-R-Sullivan commented:
@lboxell - I think that this is ready for another check. I moved the size warning code to its own script, so the release script will no longer issue size warnings. I propose adding a note to the template outline to incorporate the size-warning functionality into the template once we release version 4.0.0.

lboxell commented:
@M-R-Sullivan,

There is a dev-four template [branch](https://github.com/gslab-econ/template/tree/dev_four) that we can update to include this functionality. The goal is to have the template and gslab_python dev_four branches in-sync before pinging the PIs for review in #58. 

(I'll review your changes shortly)

M-R-Sullivan commented:
@lboxell - OK. In that case, I'll make a size-warning implementation branch off that branch.

lboxell commented on gslab_scons/tests/test_size_warning.py:
```diff
@@ -0,0 +1,312 @@
+
+        self.assertEqual(len(sizes), 3)
+
+        # Check that test.txt and test.jpg are in the dictionary
```
Update comment.

lboxell commented on gslab_scons/tests/test_size_warning.py:
```diff
@@ -0,0 +1,312 @@
+        # Check that test.txt and test.jpg are in the dictionary
+        root_path = [k for k in sizes.keys() if re.search('root_file.txt$', k)]
+        pdf_path  = [k for k in sizes.keys() if re.search('test.pdf$', k)]
+        self.assertTrue(bool(root_path))
```
I might prefer a length check versus a boolean, as it seems odd to do boolean on a list. Applies throughout.

M-R-Sullivan replied:
I will replace these sorts of calls with length checks where I find them.

lboxell commented on gslab_scons/tests/test_size_warning.py:
```diff
@@ -0,0 +1,312 @@
+
+    for i in range(len(roots)):
+        if re.search('^%s' % path, roots[i]):
+            yield (roots[i], directories[i], files[i])
```
Is this so `path = test_files` also returns the contents of `test_files/size_test`? A comment might be good.

M-R-Sullivan replied:
The check tries to ensure that walking `release` won't provide info about `test_files` and vice versa. Comment added.

M-R-Sullivan replied:
And, yes, I don't just check equality so `path = test_files` also yield `test_files/size_test`.

lboxell commented on gslab_scons/tests/test_size_warning.py:
```diff
@@ -0,0 +1,312 @@
+def isdir_dict_side_effect(*args, **kwargs):
+    '''
+    Mock os.path.isdir() so that it only recognises ./test_files/
+    as an existing directory.
```
Update comment.

lboxell commented on gslab_scons/tests/test_size_warning.py:
```diff
@@ -0,0 +1,312 @@
+
+#== Side effects for testing create_size_dictionary() ===
+# These functions mock a directory containing files of various sizes
+# and a system that recognises no directory other than that one.
```
Update comment.

lboxell commented on gslab_scons/tests/test_size_warning.py:
```diff
@@ -0,0 +1,312 @@
+
+
+def walk_dict_side_effect(*args, **kwargs):
+    '''Mock os.walk() for a mock directory called ./test_files/.'''
```
Update comment.

lboxell commented on gslab_scons/tests/test_size_warning.py:
```diff
@@ -0,0 +1,312 @@
+def getsize_dict_side_effect(*args, **kwargs):
+    '''
+    Mock os.path.getsize() to return mock sizes of files in our
+    mock ./test_files/ directory.
```
Update comment.

lboxell commented on gslab_scons/tests/test_size_warning.py:
```diff
@@ -0,0 +1,312 @@
+
+#== Side effects for testing list_ignored_files() ===
+# Define the mock file structure for testing list_ignored_files()
+struct = {'.': ['untracked.txt', 'make.log', 'make.py'],
```
Is the '.' ever used?

M-R-Sullivan commented on gslab_scons/tests/test_size_warning.py:
```diff
@@ -0,0 +1,312 @@
+    if path not in struct.keys():
+        raise StopIteration
+
+    if path == 'release':
```
You're right. I never caught this because it doesn't matter for the mock in question (we never walk `.`) but I'll correct it now.

M-R-Sullivan commented on gslab_scons/size_warning.py:
```diff
@@ -0,0 +1,111 @@
+    for line in message[ignored_line:len(message)]:
+        as_path = os.path.normpath(line)
+
+        # Only check for ignored files in the specified "look_in" directories
```
I just wasn't sure that we'd want to wait  to list obviously unversioned files (below is for `congress_text`).
```py
>>> import gslab_scons
>>> import time
>>> t = time.time(); build_files = gslab_scons.size_warning.list_ignored_files('build'); elapsed = time.time() - t
>>> print elapsed
42.23977017402649
```
Forty-two seconds isn't too long, so I'm happy to reconsider if you'd like.

M-R-Sullivan commented:
@lboxell - I made some additional changes since your last review. Namely, I added a SCons builder that runs the state-of-repo logger and `issue_size_warnings()`. I also updated the tests, and took the user input and `sys.exit()` bits out of the size warnings. 

Two notes:
- I can remove the `look_in` options if you'd like.
- I could also add an option, as a part of this issue or in another issue, that allows the user to turn off warnings for certain files (i.e. so warnings aren't issued after every SCons run when the user doesn't care that a file is too large).

lboxell commented on gslab_scons/tests/test_size_warning.py:
```diff
@@ -228,12 +278,28 @@ def walk_ignored_side_effect(*args, **kwargs):
         raise StopIteration
 
-    if path == 'release':
+    if path == '.':
```
Are these 'subdirs' used anymore?

M-R-Sullivan replied:
Nay. I've deleted them.

lboxell commented on gslab_scons/tests/test_size_warning.py:
```diff
@@ -260,38 +326,40 @@ def isfile_ignored_side_effect(*args, **kwargs):
+    return os.path.normpath(path) in acceptable
 
 
 def walk_dict_side_effect(*args, **kwargs):
```
Seems like we have two pretty similar walk side effects. If its easy to combine, we should. If not, not a big deal.

M-R-Sullivan replied:
I combined them. I think the mocking in this script could be further polished, although I'm unsure whether it's worth spending time to make improvements at this point.

lboxell commented on gslab_scons/tests/test_size_warning.py:
```diff
@@ -88,24 +121,36 @@ def test_list_ignored_files(self, mock_check, mock_walk,
+                          'release/subdir/ignored.txt']
+
+        self.assertEqual(len(ignored), len(expect_ignored))
+        self.assertEqual(ignored[0], expect_ignored[0])          
```
I like the `for i in range(len(ignored)):` check that we do in lines 132-133 a bit better.

M-R-Sullivan replied:
OK!

lboxell commented:
Lets add your point 2, to the outline and see if we feel like it would be useful after we have some more experience.

M-R-Sullivan commented on gslab_scons/size_warning.py:
```diff
@@ -89,6 +78,16 @@ def list_ignored_files(look_in):
     return ignore_files
 
 
+def _is_subpath(inner, outer):
```
I couldn't find any standard way to do this in Python in my brief search.

M-R-Sullivan commented on gslab_scons/size_warning.py:
```diff
@@ -20,24 +20,18 @@ def issue_size_warnings(look_in = ['source', 'raw', 'release'],
 
         if size > limit and file_name:
-            print "Warning: the versioned file %s is larger than %d MB.\n" \
+            print "Warning: the versioned file %s is larger than %f MB." \
```
Agreed!

M-R-Sullivan commented on gslab_scons/misc.py:
```diff
@@ -4,13 +4,28 @@
+    state_of_repo(maxit)
+
+    # Issue size warnings
+    look_in = env['look_in']
```
I vote for keeping it explicit for now.

M-R-Sullivan commented:
@lboxell - I think I've responded to each of your comments and added my second point to the [outline](https://github.com/gslab-econ/gslab_python/wiki/Outline). I also added a little function for making the "Warning" text red and bold (I can get rid of this if you'd like). Let me know what you think.

lboxell commented:
@M-R-Sullivan, Looks good!

M-R-Sullivan commented:
Great. Merging now.
